### ISSUE TITLE
RuntimeError: You are trying to run nbclient in an environment where an event loop is already running. Please pass `nest_asyncio=True` in `NotebookClient.execute` and such methods.

### ISSUE THREAD

**Issue by linneuholanda** [nteract/papermill#490]
When running papermill in a conda enviroment, I am getting the following traceback:

```python
RuntimeError                              Traceback (most recent call last)
~/anaconda3/envs/ml_top_phases/lib/python3.7/site-packages/nbclient/util.py in wrapped(self, *args, **kwargs)
     36         try:
---> 37             result = loop.run_until_complete(coro(self, *args, **kwargs))
     38         except RuntimeError as e:

~/anaconda3/envs/ml_top_phases/lib/python3.7/asyncio/base_events.py in run_until_complete(self, future)
    562         self._check_closed()
--> 563         self._check_runnung()
    564 

~/anaconda3/envs/ml_top_phases/lib/python3.7/asyncio/base_events.py in _check_runnung(self)
    522         if self.is_running():
--> 523             raise RuntimeError('This event loop is already running')
    524         if events._get_running_loop() is not None:

RuntimeError: This event loop is already running

During handling of the above exception, another exception occurred:

RuntimeError                              Traceback (most recent call last)
<ipython-input-3-da6a15f1c881> in <module>
      5                     output_file,
      6                     parameters=parameters,
----> 7                     kernel_name=kernel_name)

~/anaconda3/envs/ml_top_phases/lib/python3.7/site-packages/papermill/execute.py in execute_notebook(input_path, output_path, parameters, engine_name, request_save_on_cell_execute, prepare_only, kernel_name, progress_bar, log_output, stdout_file, stderr_file, start_timeout, report_mode, cwd, **engine_kwargs)
    102                     stdout_file=stdout_file,
    103                     stderr_file=stderr_file,
--> 104                     **engine_kwargs
    105                 )
    106 

~/anaconda3/envs/ml_top_phases/lib/python3.7/site-packages/papermill/engines.py in execute_notebook_with_engine(self, engine_name, nb, kernel_name, **kwargs)
     47     def execute_notebook_with_engine(self, engine_name, nb, kernel_name, **kwargs):
     48         """Fetch a named engine and execute the nb object against it."""
---> 49         return self.get_engine(engine_name).execute_notebook(nb, kernel_name, **kwargs)
     50 
     51 

~/anaconda3/envs/ml_top_phases/lib/python3.7/site-packages/papermill/engines.py in execute_notebook(cls, nb, kernel_name, output_path, progress_bar, log_output, autosave_cell_every, **kwargs)
    341         nb_man.notebook_start()
    342         try:
--> 343             cls.execute_managed_notebook(nb_man, kernel_name, log_output=log_output, **kwargs)
    344         finally:
    345             nb_man.cleanup_pbar()

~/anaconda3/envs/ml_top_phases/lib/python3.7/site-packages/papermill/engines.py in execute_managed_notebook(cls, nb_man, kernel_name, log_output, stdout_file, stderr_file, start_timeout, execution_timeout, **kwargs)
    400             stderr_file=stderr_file,
    401         )
--> 402         return PapermillNotebookClient(nb_man, **final_kwargs).execute()
    403 
    404 

~/anaconda3/envs/ml_top_phases/lib/python3.7/site-packages/papermill/clientwrap.py in execute(self, **kwargs)
     34         self.reset_execution_trackers()
     35 
---> 36         with self.setup_kernel(**kwargs):
     37             self.log.info("Executing notebook with kernel: %s" % self.kernel_name)
     38             self.papermill_execute_cells()

~/anaconda3/envs/ml_top_phases/lib/python3.7/contextlib.py in __enter__(self)
    110         del self.args, self.kwds, self.func
    111         try:
--> 112             return next(self.gen)
    113         except StopIteration:
    114             raise RuntimeError("generator didn't yield") from None

~/anaconda3/envs/ml_top_phases/lib/python3.7/site-packages/nbclient/client.py in setup_kernel(self, **kwargs)
    402 
    403         if not self.km.has_kernel:
--> 404             self.start_new_kernel_client(**kwargs)
    405         try:
    406             yield

~/anaconda3/envs/ml_top_phases/lib/python3.7/site-packages/nbclient/util.py in wrapped(self, *args, **kwargs)
     39             if str(e) == 'This event loop is already running':
     40                 raise RuntimeError(
---> 41                     'You are trying to run nbclient in an environment where an '
     42                     'event loop is already running. Please pass `nest_asyncio=True` in '
     43                     '`NotebookClient.execute` and such methods.'

RuntimeError: You are trying to run nbclient in an environment where an event loop is already running. Please pass `nest_asyncio=True` in `NotebookClient.execute` and such methods.
```
I have no idea where this comes from. Have you run into this before? The code that generates the error is written below.

```python
template = "preprocessing_template.ipynb"
parameters = {"allowed_windings":[0,1], "epsilon":0.01, "csv_dir":"/home/rio/ssh_csvs/ssh1"}
#kernel_name = "ml_top_phases"
kernel_name = "ml_top_phases"
parameters["grid_folder"]="/home/linneu/ssh_grids/ssh1/periodic_100_6561"
parameters["csv_name"]="periodic_100_6561.csv"
output_file = "preprocessing_output_ssh1_periodic_100_6561.ipynb"
pm.execute_notebook(template,
                    output_file,
                    parameters=parameters,
                    kernel_name=kernel_name)
```

**Comment by MSeal** [comment]
So it looks like you're calling papermill from an async loop. Try passing `nest_asyncio=True` as an argument to `pm.execute_notebook` and it will pass the flag to reuse the existing event loop, though the call to papermill will still be a blocking call.

**Comment by evaranouar** [comment]
Hi, 
thanks for answer, I did run through the same issue and added `nest_asyncio=True`.
I have another issue now (p.s.: the notebook I have is intended to output a csv file and the csv is there).
`TypeError: object NoneType can't be used in 'await' expression`

please find below the traceback:

`TypeError                                 Traceback (most recent call last)
<ipython-input-23-8d7848ec6b74> in <module>
      8     r'C:\Users\Anouar\workplace\ipynbEvarInvest\Stars\papermill_output.ipynb',
      9     parameters = dict( file_name = 'another_csv_from_papermill_notebook'),
---> 10     nest_asyncio=True
     11 )

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\papermill\execute.py in execute_notebook(input_path, output_path, parameters, engine_name, request_save_on_cell_execute, prepare_only, kernel_name, progress_bar, log_output, stdout_file, stderr_file, start_timeout, report_mode, cwd, **engine_kwargs)
    102                     stdout_file=stdout_file,
    103                     stderr_file=stderr_file,
--> 104                     **engine_kwargs
    105                 )
    106 

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\papermill\engines.py in execute_notebook_with_engine(self, engine_name, nb, kernel_name, **kwargs)
     47     def execute_notebook_with_engine(self, engine_name, nb, kernel_name, **kwargs):
     48         """Fetch a named engine and execute the nb object against it."""
---> 49         return self.get_engine(engine_name).execute_notebook(nb, kernel_name, **kwargs)
     50 
     51 

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\papermill\engines.py in execute_notebook(cls, nb, kernel_name, output_path, progress_bar, log_output, autosave_cell_every, **kwargs)
    341         nb_man.notebook_start()
    342         try:
--> 343             cls.execute_managed_notebook(nb_man, kernel_name, log_output=log_output, **kwargs)
    344         finally:
    345             nb_man.cleanup_pbar()

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\papermill\engines.py in execute_managed_notebook(cls, nb_man, kernel_name, log_output, stdout_file, stderr_file, start_timeout, execution_timeout, **kwargs)
    400             stderr_file=stderr_file,
    401         )
--> 402         return PapermillNotebookClient(nb_man, **final_kwargs).execute()
    403 
    404 

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\papermill\clientwrap.py in execute(self, **kwargs)
     39             info_msg = self._wait_for_reply(self.kc.kernel_info())
     40             self.nb.metadata['language_info'] = info_msg['content']['language_info']
---> 41             self.set_widgets_metadata()
     42 
     43         return self.nb

~\AppData\Local\anaconda3\envs\papermill_env\lib\contextlib.py in __exit__(self, type, value, traceback)
    117         if type is None:
    118             try:
--> 119                 next(self.gen)
    120             except StopIteration:
    121                 return False

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\nbclient\client.py in setup_kernel(self, **kwargs)
    406             yield
    407         finally:
--> 408             self._cleanup_kernel()
    409 
    410     @asynccontextmanager

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\nbclient\util.py in wrapped(self, *args, **kwargs)
     35             nest_asyncio.apply(loop)
     36         try:
---> 37             result = loop.run_until_complete(coro(self, *args, **kwargs))
     38         except RuntimeError as e:
     39             if str(e) == 'This event loop is already running':

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\nest_asyncio.py in run_until_complete(self, future)
     57         if not f.done():
     58             raise RuntimeError('Event loop stopped before Future completed.')
---> 59         return f.result()
     60 
     61     def _run_once(self):

~\AppData\Local\anaconda3\envs\papermill_env\lib\asyncio\futures.py in result(self)
    179         self.__log_traceback = False
    180         if self._exception is not None:
--> 181             raise self._exception
    182         return self._result
    183 

~\AppData\Local\anaconda3\envs\papermill_env\lib\asyncio\tasks.py in __step(***failed resolving arguments***)
    247                 # We use the `send` method directly, because coroutines
    248                 # don't have `__iter__` and `__next__` methods.
--> 249                 result = coro.send(None)
    250             else:
    251                 result = coro.throw(exc)

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\nbclient\client.py in _async_cleanup_kernel(self)
    338                 # shutdowns have not been successful or called yet, so give a direct kill
    339                 # call here and recover gracefully if it's already dead.
--> 340                 await ensure_async(self.km.shutdown_kernel(now=True))
    341             except RuntimeError as e:
    342                 # The error isn't specialized, so we have to check the message

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\nbclient\util.py in ensure_async(obj)
     55     if inspect.isawaitable(obj):
     56         try:
---> 57             result = await obj
     58         except RuntimeError as e:
     59             if str(e) == 'cannot reuse already awaited coroutine':

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\jupyter_client\manager.py in shutdown_kernel(self, now, restart)
    583 
    584         if now:
--> 585             await self._kill_kernel()
    586         else:
    587             self.request_shutdown(restart=restart)

~\AppData\Local\anaconda3\envs\papermill_env\lib\site-packages\jupyter_client\manager.py in _kill_kernel(self)
    645                     await self.signal_kernel(signal.SIGKILL)
    646                 else:
--> 647                     await self.kernel.kill()
    648             except OSError as e:
    649                 # In Windows, we will get an Access Denied error if the process

TypeError: object NoneType can't be used in 'await' expression`

**Comment by MSeal** [comment]
Your installed `jupyter_client` is old, upgrade it to the latest (`6.1.2` atm) and that error should go away. It's probably a good idea to start with a new environment or upgrade everything jupyter related.

**Comment by evaranouar** [comment]
upgrading `jupyter_client` to `6.1.2` nailed it, worked for me.
many thanks

**Comment by mirekphd** [comment]
I got the OP's original `RuntimeError` (raised despite new parameter values being correctly injected to the output notebook) when invoking `execute_notebook()` in a containerized python container with the latest jupyter notebook installation modeled on the official [jupyter/base-notebook](https://hub.docker.com/r/jupyter/base-notebook/), with the latest versions of relevant packages. 

@MSeal's suggestion of setting the undocumented parameter `nest_asyncio` to True worked as advertised. May I suggest that this parameter becomes documented and initialized with a sane default (of True)?

**Comment by MSeal** [comment]
@davidbrochart What would you recommend we do for `nest_asyncio` here? You have more experience with python's async model than I do and would love your input on how we avoid confusing users as changing this default I believe makes other patterns then fail by default.

**Comment by davidbrochart** [comment]
Just to summarize why we introduced this parameter: nbclient is a relatively new library responsible for executing notebooks, and I suppose papermill now uses it. As the Jupyter ecosystem is moving towards an asynchronous execution model (using asyncio's coroutines instead of plain blocking functions), we started introducing async execution into nbclient. In order to have one single code base (and not duplicate the code between blocking and async functions), the heart of nbclient is async, and to keep compatibility with the blocking API, each blocking function is a wrapper around the async function, which runs the coroutine until completion in the event loop. The problem is that if you are already running in an async environment, the event loop is already running, so you cannot just instantiate another event loop, you have to run everything in the same event loop. The nest_asyncio library allows to do that, by nesting one event loop into another.
The reason for having `nest_asyncio=False` by default is that nest_asyncio doesn't play well with e.g. Tornado (see https://github.com/tornadoweb/tornado/issues/2753).
That being said, nest_asyncio is not the only solution. We could also run nbclient's event loop in its own thread (only for the blocking API). That would make each event loop completely separated, and there would be no need for nesting event loops. If we have too many issues using nest_asyncio, we might want to do that in nbclient.

**Comment by MSeal** [comment]
The threaded event loop might be a good default behavior for external callers to fall into, however this then runs into issues with the fact that we're indirectly spawning subprocesses and threading -> multiprocessing can lead to deadlocks.

**Comment by davidbrochart** [comment]
Maybe we can experiment? I'll open a PR in nbclient.

**Comment by MSeal** [comment]
@linneuholanda This should be fixed when nbclient 0.4.0 releases. You can try out the fix on the current alpha 0.4.0a0 if you like.

**Comment by mirekphd** [comment]
It seems relevant to note that the need for the use of `nest_asyncio=True` arises only inside Jupyter Notebooks. 

If you run your script (e.g. with papermill's `pm.execute_notebook()`) outside of Jupyter notebooks python kernel, in a normal python program executed in the system shell (even in a terminal provided by Jupyter), then there is no need for nesting the event loop - all multiprocessing scripts I tested run without this error. This functionality can be added to your custom scripts (if you prefer using Jupyter) by the following commands (first `pip install [nest-asyncio](https://pypi.org/project/nest-asyncio/)` if you don't have it - you should if you have `papermill`):
```
# patch asyncio to make Jupyter's event loop re-entrant
import nest_asyncio 
nest_asyncio.apply()
```
